### DISCUSSION TITLE
【犀牛鸟 Issue #52】稀疏 ES-MoE 专家剪枝的效率-精度分析:无损剪枝、两处 prune→retrain 框架 bug 修复,与恢复路径的失效边界

### DISCUSSION THREAD

**Discussion by Lfan-ke** [Tencent/YOLO-Master#197] [Show and tell]
本文是 Issue #52(MoE 专家剪枝与动态超参数调度优化)的技术总结。完整脚本、数据、图见实验仓库 https://github.com/Lfan-ke/yolomoe-ln 。过程中发现并向上游修复的两个框架 bug 见 PR https://github.com/Tencent/YOLO-Master/pull/192 与 https://github.com/Tencent/YOLO-Master/pull/194 。

## 复现环境

<div align="center">

算力平台 · <a href="https://ai.gitee.com/">模力方舟(Gitee AI)</a>

<table>
<tr><td align="center"><b>GPU</b></td><td align="center">NVIDIA RTX 5090 · 32 GB × 1</td></tr>
<tr><td align="center"><b>CPU</b></td><td align="center">Intel Xeon Platinum 8358P</td></tr>
<tr><td align="center"><b>内存 / 磁盘</b></td><td align="center">48 GB / 系统盘 30 GB + 数据盘 50 GB</td></tr>
<tr><td align="center"><b>OS / Python / CUDA</b></td><td align="center">Ubuntu 24.04.1 LTS · Python 3.12.3 · CUDA 12.8</td></tr>
<tr><td align="center"><b>PyTorch / ultralytics</b></td><td align="center">2.6.0a0+nv25.01 (NGC) / 8.4.101</td></tr>
<tr><td align="center"><b>optuna / numpy / opencv / filelock</b></td><td align="center">4.9.0 / 1.26.4 / 4.10.0 / 3.15.4</td></tr>
</table>

</div>

## 摘要

在稀疏 ES-MoE(num_experts=4, top_k=2)的 YOLO-Master-EsMoE-N 上,专家利用率高度不均衡:每个 ES_MOE 块的 4 个专家中仅 2 个被稳定激活。MoEPruner 按利用率剪除低利用专家(4→2),direct 推理下 mAP50 零损失(0.4271→0.4269)、推理延迟降 24.5%、参数降 9.2%。issue 要求的"剪枝后恢复"步骤在本无损场景不必要;实现该步骤时定位并修复了 prune→retrain 路径上的两个框架 bug(专家数 re-inflation、专家核大小丢失),二者均使重训从随机初始化的专家起步。修复后重训机制正确:全量微调回升至基线约 97%(mAP50 0.413),但对已无损的剪枝模型 LoRA 微调不提升精度(反因训练态统计漂移降至 0.233)。动态 top_k 退火调度对训练收敛无影响(与基线曲线逐点相同),其价值在推理侧稀疏而非收敛加速。

## 1. 基线与专家利用率

YOLO-Master-EsMoE-N(4 个 ES_MOE 块,每块 num_experts=4 / top_k=2)在 VisDrone(train 6471 / val 548,10 类)imgsz 1024 训练至收敛(100 epoch):

- mAP50 = 0.4271,mAP50-95 = 0.2524(达优秀标准 ≥0.40)
- FLOPs = 26.58 G,Latency = 17.06 ms/图(std 0.12),Params = 2.814 M

指标口径:mAP 用 `YOLO.val()`;FLOPs 用 `torch.profiler` 真实 val 图前向计数;Latency 单图 100 次均值 + 20 次预热;专家利用率经 `model.val()` + 路由钩子统计,按每层 `num_experts` 补齐未激活专家后计 Gini。

专家利用率 Gini 系数定义(每层 usage 向量 $u=(u_1,\dots,u_E)$,$E$ 为专家数):

$$G = \frac{\sum_{i=1}^{E}\sum_{j=1}^{E} |u_i - u_j|}{2 E \sum_{i=1}^{E} u_i}$$

实测每层利用率:

| 层 | 每专家利用率 | 激活数 | Gini |
|---|---|---|---|
| model.3 | 0.00 / 0.50 / 0.00 / 0.50 | 2 | 0.50 |
| model.6 | 0.50 / 0.50 / 0.00 / 0.00 | 2 | 0.50 |
| model.9 | 0.00 / 0.50 / 0.50 / 0.00 | 2 | 0.50 |
| model.12 | 0.10 / 0.11 / 0.46 / 0.33 | 4 | 0.32 |

均值 Gini = 0.455。前三层恰有 2 个专家零利用(路由从不选中),第四层四专家均被使用但两个占比低(0.10 / 0.11)。负载均衡损失未能避免这种崩溃,这与官方分析(issue #36 routing collapse)一致,也为剪枝提供了真实冗余目标。

## 2. 剪枝(direct 推理):无损效率结果

用 MoEPruner 按利用率阈值 {0.05, 0.10, 0.15, 0.20, 0.30} 剪枝,全阈值一致地把每层 4 专家剪到 2(移除低利用专家),direct 推理(不重训):

| 阈值 | mAP50 | mAP50-95 | FLOPs(G) | Latency(ms) | Params(M) | Gini |
|---|---|---|---|---|---|---|
| 基线(4 专家) | 0.4271 | 0.2524 | 26.58 | 17.06 | 2.814 | 0.455 |
| 0.05 / 0.10 | 0.4271 | 0.2524 | 26.59 | 13.20 | 2.697 | 0.080 |
| 0.15 / 0.20 / 0.30 | 0.4269 | 0.2528 | 26.61 | 12.87 | 2.556 | 0.0 |

mAP50 零损失(0.4271→0.4269),延迟 -24.5%(17.06→12.87 ms),参数 -9.2%(2.814→2.556 M),Gini 0.455→0(剪除死专家后剩余专家均衡)。FLOPs 几乎不变:top_k=2 的稀疏推理本就不计算未选中的专家,故理论 FLOPs 不因剪枝下降;但物理剪除消除了 4 选 2 的路由与 dispatch 开销,真实延迟显著下降。这是 MoEPruner 的正确用例:清理崩溃的冗余专家,免费换取推理加速。精度-延迟 Pareto 与阈值曲线见仓库 `results/figures/`,sweet spot 取阈值 0.30。

## 3. prune→retrain 路径的两个框架 bug

issue 要求对比"剪枝后 direct 推理 vs 剪枝后微调恢复"。实现微调恢复时发现:`YOLO(pruned.pt).train()` 得到的模型在第 1 个 epoch 的 val mAP50 仅约 0.002,而同一权重文件 direct 推理为 0.427。逐层比对权重加载后定位到两处 bug,均使重训从随机初始化的专家起步。

训练加载路径(`setup_model` → `load_checkpoint` → 取 `weights.yaml` → `get_model(cfg=yaml, weights)`)从 YAML 重建模型再以 `intersect_dicts` 按名与形状匹配加载权重:

```mermaid
flowchart TD
    A["YOLO(pruned.pt).train()"] --> B["从 model.yaml 重建 DetectionModel"]
    B --> C{"YAML 是否编码剪枝架构?"}
    C -->|"bug1: 只有默认 num_experts"| D["重建 4 专家 → 剪枝的 2 专家权重形状不匹配被丢弃"]
    C -->|"bug2: 未编码保留专家的异构核大小"| E["重建默认核 [3,5] → 保留专家实际核 [5,9] 不匹配被丢弃"]
    D --> F["intersect_dicts 静默丢弃 → 专家随机初始化 → 重训起点崩坏"]
    E --> F
    F --> G["修复: YAML 写入完整 ES_MOE 参数(专家数 + per-expert 核大小)"]
    G --> H["重建精确复刻剪枝架构 → 权重全部加载 → 重训起点正确"]
```

- bug 1(PR #192,已合入):YAML 只记录了通道数,未记录剪枝后的 num_experts,重建用 ES_MOE 默认专家数,剪枝的专家/路由权重因形状不匹配被 `intersect_dicts` 丢弃,模型被重新膨胀回满专家。修复:剪枝后把 num_experts 写入 YAML。
- bug 2(PR #194):MoEPruner 保留的是利用率最高的专家,其深度卷积核大小异构(实测各层保留核为 [5,9] / [3,5] / [5,7] / [7,9]);而 ES_MOE 从 YAML 重建时按默认 `[3,5,7][:num_experts]` 赋核,3/4 层核大小不匹配,对应深度卷积权重被丢弃。修复:ES_MOE 增加可选 `expert_kernel_sizes` 参数,MoEPruner 把保留专家的核大小写入 YAML。二者均向后兼容,附回归测试。

## 4. 恢复策略(修复后):direct vs LoRA vs 全量微调

以修复后的剪枝模型(权重正确加载,重训起点 val mAP50 ≈ 0.24,train loss 从约 1.7 起而非崩坏的 3.2)对比三种策略,全部 imgsz 1024:

| 策略 | mAP50 | 说明 |
|---|---|---|
| direct(不重训) | 0.4269 | 剪枝零损失,恢复其实不必要 |
| 全量微调(8 ep) | 0.413 | 全参可训,回升至基线约 97%,仍略低于 direct |
| LoRA 微调(10 ep) | 0.2331 | 冻结基座 + 适配器;lr∈{1e-4,3e-4,1e-3} 结果一致 |

全量微调随 epoch 稳定回升(ep5→ep7:0.403 → 0.405 → 0.413),达基线 mAP50 0.4271 的约 97%,与学习资料"恢复可达全量 95-98%"一致;但仍略低于 direct 的 0.4269——剪枝本身无损,重训无缺口可补,反而受训练态增广与 BatchNorm 重估轻微拉低。LoRA 恢复则在多组超参下一致收敛到 mAP50 = 0.2331(与 lora_r / lr / weight_decay 无关):冻结基座使模型无法通过少量适配器抵消 BatchNorm 统计漂移,故显著低于全量微调与 direct。结论:全量微调可恢复至近基线,但在本无损剪枝场景 direct 推理已是最优,微调恢复不必要;LoRA 在此场景不适用。

## 5. 动态调度与收敛加速比

实现 DynamicTopKScheduler:按验证 mAP 饱和与专家利用率均衡(Gini)双信号退火 top_k。判据(每验证 epoch,窗口 $w$):

$$\text{saturated}_t = \max(\text{mAP}_{t-w+1:t}) - \max(\text{mAP}_{t-2w+1:t-w}) < \varepsilon,\quad \text{balanced}_t = \overline{G}_t \le G_{\text{ceil}}$$

连续 `hold` 个 epoch 满足且 $k>k_{\min}$ 时 $k \leftarrow k-1$;退火后 mAP 跌幅超阈值则回滚 $k \leftarrow k+1$ 并冻结。收敛加速比 = 基线达到其最终 mAP 的 95% 所需 epoch / 实验组所需 epoch。

固定基线组与动态退火组对照,稀疏 imgsz 1024,30 epoch,固定随机种子:

**动态退火组与基线组的 mAP 曲线逐点完全相同**(如 epoch 25/26/27 两组均为 0.13217 / 0.13382 / 0.13611),收敛加速比 = 1.00。机制:ES_MOE 训练前向走 dense 路径(`use_dense = self.training`),top_k 只作用于推理稀疏路由、不影响训练;退火在精度饱和后才触发,对训练收敛无贡献。去 Gini 门的消融组由同一机制必然给出相同曲线。这是如实的 null 结果——top_k 退火的价值在推理侧 FLOPs 缩减,而非训练收敛加速。若要影响收敛,应调度进入损失的量(如 balance-loss 系数),而非仅作用于推理的 top_k。

## 6. 场景化部署建议

issue 示例假设"server 用低阈值、edge 用高阈值"以取舍精度与效率。实测本稀疏模型全阈值 {0.05-0.30} 一致剪到每层 2 专家、精度全程无损,故阈值在此不构成精度-效率旋钮:

- server 与 edge 统一推荐阈值 0.15-0.30,得 -24.5% 延迟、-9.2% 参数、零精度损失,无需按部署端调阈值。
- 真正的部署旋钮是稀疏架构本身(num_experts / top_k),而非剪枝阈值;更激进的边缘压缩应从架构层降低专家数或 top_k。
- 剪枝前先测专家利用率:存在崩溃/零利用专家(Gini 高)时剪枝是免费效率;专家均衡无冗余则无可剪。

## 7. 结论与可复现

- 稀疏 ES-MoE 的专家剪枝是真实可用的效率手段:清理崩溃专家,-24.5% 延迟、-9.2% 参数、mAP 零损失、无需重训。
- prune→retrain 路径存在两处会使重训崩坏的框架 bug,已定位、修复并向上游提交 PR #192(已合入)与 #194(含回归测试)。
- 无损剪枝场景下,微调恢复不必要:全量微调可回升至基线约 97% 但仍低于 direct,LoRA 微调不提升精度。
- 动态 top_k 退火不加速训练收敛(架构上只影响推理),其价值在推理效率。

全部脚本、数据、图开源可复现:https://github.com/Lfan-ke/yolomoe-ln 。框架 bug 修复 PR:#192、#194。动态调度 PR:#183。
